server/src/controller/PlaceController.hh

```diff
@@ -0,0 +1,249 @@
+        return createDtoResponse(Status::CODE_200, service_.getOne(id));
+    }
+
+    ENDPOINT_INFO(search)
+    {
+        info->summary = "Search places";
+        info->tags.emplace_back("place-controller");
+
+        info->queryParams["query"].required = false;
+        info->queryParams["limit"].required = false;
+        info->queryParams["offset"].required = false;
+
+        info->addResponse<Object<dto::PlacePageDto>>(Status::CODE_200,
+                                                     "application/json");
+        info->addResponse<Object<dto::StatusDto>>(Status::CODE_404,
+                                                  "application/json");
+        info->addResponse<Object<dto::StatusDto>>(Status::CODE_500,
+                                                  "application/json");
+    }
+
+    ENDPOINT("GET",
+             "place",
+             search,
+             QUERY(String, query, "query", std::string{}),
+             QUERY(UInt64, limit, "limit", 20U),
+             QUERY(UInt64, offset, "offset", 0UL))
+    {
+        return createDtoResponse(Status::CODE_200,
+                                 service_.search(query, limit, offset));
+    }
+
+    ENDPOINT_INFO(putOne)
+    {
+        info->summary = "Put one place";
+        info->tags.emplace_back("place-controller");
+        info->addSecurityRequirement("JWT Bearer Auth", {});
+
+        info->addResponse<Object<dto::PlaceDto>>(Status::CODE_200,
+                                                 "application/json");
+        info->addResponse<Object<dto::StatusDto>>(Status::CODE_400,
+                                                  "application/json");
+        info->addResponse<Object<dto::StatusDto>>(Status::CODE_401,
+                                                  "application/json");
+        info->addResponse<Object<dto::StatusDto>>(Status::CODE_403,
+                                                  "application/json");
+        info->addResponse<Object<dto::StatusDto>>(Status::CODE_500,
+                                                  "application/json");
+    }
+
+    ENDPOINT("PUT",
+             "place",
+             putOne,
+             AUTHORIZATION(std::shared_ptr<JWT::Payload>, auth_object),
+             BODY_DTO(Object<dto::PlaceDto>, dto))
+    {
+        OATPP_ASSERT_HTTP(dto->id && dto->owner_id && dto->address &&
+                            dto->latitude && dto->longitude,
+                          Status::CODE_400,
+                          "Required parameter not provided")
+
+        try {
+            OATPP_ASSERT_HTTP(
+              auth_object->role == 0 ||
+                (((dto->owner_id == nullptr &&
+                   (dto->owner_id = auth_object->user_id)) && // NOLINT
+                  auth_object->user_id == service_.getOne(dto->id)->owner_id) ||
+                 dto->owner_id == auth_object->user_id),
+              Status::CODE_403,
+              "Cannot create or modify place of another user as a regular user")
```

Comment:
This is unreadable. Also the bug from previous conversation still exists.

Reply:
Isn't it possible for a user to PUT a dto with NULL owner_id? Or does oatpp require that all of the dto's fields are non-null.

Reply:
it shouldn't contain that bug as if the `dto->owner_id == nullptr` then `dto->owner_id = auth_object->user_id` (notice single `=`) and `auth_object->user_id` is not null bc of `AUTHORIZATION` macro

Reply:
Oh maybe you meant to assert `dto->owner_id == auth_object->user_id` before using that `getOne` method? Then it would make sense

Reply:
> it shouldn't contain that bug as if the `dto->owner_id == nullptr` then `dto->owner_id = auth_object->user_id` (notice single `=`) and `auth_object->user_id` is not null bc of `AUTHORIZATION` macro

But if an administrator makes that request `auth_object->role == 0`, then the later `dto->owner_id = auth_object->user_id` is not evaluated. And afaik `dto->owner_id` can be NULL.

As to the previous bug. It existed because `dto->owner_id == auth_object->user_id` was evaluated after `service_.getOne()`, and that is still the case.

Reply:
Okay, I didn't notice the assert few lines above. So I see we *do* make sure that all of the dto fields are non-NULL. That means that `(dto->owner_id == nullptr && (dto->owner_id = auth_object->user_id))` is redundant, because we know that `dto->owner_id` has to contain a value.

And because here `dto->owner_id == nullptr` is always false, `service_.getOne()` will never be called, so the bug does not exist anymore.

Doesn't this thing simplify to just `auth_object->role == 0 || dto->owner_id == auth_object->user_id`?

Reply:
> `dto->owner_id` has to contain a value

that's not true, checks above only check if all fields which are marked as `NOT NULL` in db are not null. `owner_id` is a special case which we check in the second macro

Reply:
> > `dto->owner_id` has to contain a value
> 
> that's not true, checks above only check if all fields which are marked as `NOT NULL` in db are not null. `owner_id` is a special case which we check in the second macro

To clarify I was talking about the asserts on the begging of the function:
```c++
OATPP_ASSERT_HTTP(dto->id && dto->owner_id && dto->address && dto->latitude && dto->longitude, Status::CODE_400, "Required parameter not provided")
```

So `dto->owner_id` *has* to contain a value.
And thus the whole:
```c++
((dto->owner_id == nullptr && (dto->owner_id = auth_object->user_id)) && auth_object->user_id == service_.getOne(dto->id)->owner_id)
```
is rendered redundant, because `dto->owner_id == nullptr` can't evaluate to `true`.

Reply:
Maybe `dto->owner_id` wasn't supposed to be included in that assert?

Reply:
You're right, I must have poor eyesight because I overlooked that even though I was looking for it

Reply:
On my second thought I think, that we should not assign anything to DTO in controllers. I'll remake this

Reply:
And also `dto->id` can be null bc of `PRIMARY KEY` autoincrement in db

Reply:
Okay, nvm, it can be null in `createOne()` method, but it shouldn't be so in `replaceOne()`

Reply:
solved in a66ebab2994dbf1c60d2e6cd9538fa76c7e6b23e